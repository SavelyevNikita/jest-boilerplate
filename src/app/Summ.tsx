
import * as React from 'react';

type SummType = React.ReactElement<SummProps>

interface SummProps {
    children?: SummType | SummType[];
    v: number;
}

export const Summ =
(props: SummProps): SummType => {
    const k = props.v;

    if (Array.isArray(props.children)) {
        return (
            <div>
                { props.children.map((child: SummType, key: number): SummType =>
                    <Summ key={key.toString()} v={k + child.props.v}>
                        { child.props.children }
                    </Summ>)}
            </div>
        );
    }
    const i = k + (props.children ? props.children.props.v : 0);
    return (
        <div>
            {props.children ?
                <Summ v={i}>
                    { props.children.props.children }
                </Summ>
                : k}
        </div>
    );
}
