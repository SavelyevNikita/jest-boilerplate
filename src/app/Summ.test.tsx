
import * as React from 'react';
import { mount } from 'enzyme';
import { Summ } from './Summ';

test('Summ exists', (): void => {
    const el = (<Summ v={0}></Summ>);
    const sh = mount(el);
    expect(sh).toMatchSnapshot();
});

test('Summ accepts one parameter', (): void => {
    const v = 2;
    const el = (<Summ v={v}></Summ>);
    const sh = mount(el);
    expect(sh.text()).toBe(v.toString());
    expect(sh).toMatchSnapshot();
});

test('Summ the parent with its soley child', (): void => {
    const v1 = 2;
    const v2 = 6;
    const expected = v1 + Number(v2);
    const el = (
        <Summ v={v1}>
            <Summ v={v2}></Summ>
        </Summ>
    );
    const sh = mount(el);
    expect(sh.text()).toBe(expected.toString());
    expect(sh).toMatchSnapshot();
});

test('Summ te parent with its two nested children', (): void => {
    const v1 = 2;
    const v2 = 6;
    const v3 = 3;
    const expected = Number(v1) + Number(v2) + Number(v3);
    const el = (
        <Summ v={v1}>
            <Summ v={v2}>
                <Summ v={v3}></Summ>
            </Summ>
        </Summ>
    );
    const sh = mount(el);
    expect(sh.text()).toBe(expected.toString());
    expect(sh).toMatchSnapshot();
});

test('Summ the parent with its two children separately', (): void => {
    const v1 = 1;
    const v2 = 2;
    const v3 = 3;
    const expected = 34;
    const el = (
        <Summ v={v1}>
            <Summ v={v2}></Summ>
            <Summ v={v3}></Summ>
        </Summ>
    );
    const sh = mount(el);
    expect(sh.text()).toBe(expected.toString());
    expect(sh).toMatchSnapshot();
});

test('Summ the parent with a complex tree of children', (): void => {
    const v1 = 1;
    const v2 = 2;
    const v3 = 3;
    const v4 = 4;
    const v5 = 5;
    const expected = 656;
    const el = (
        <Summ v={v1}>
            <Summ v={v2}>
                <Summ v={v3}></Summ>
            </Summ>
            <Summ v={v4}></Summ>
            <Summ v={v5}></Summ>
        </Summ>
    );
    const sh = mount(el);
    expect(sh.text()).toBe(expected.toString());
    expect(sh).toMatchSnapshot();
});
