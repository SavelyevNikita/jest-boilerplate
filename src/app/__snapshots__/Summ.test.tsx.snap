// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summ accepts one parameter 1`] = `
<Component
  v={2}
>
  <div>
    2
  </div>
</Component>
`;

exports[`Summ exists 1`] = `
<Component
  v={0}
>
  <div>
    0
  </div>
</Component>
`;

exports[`Summ te parent with its two nested children 1`] = `
<Component
  v={2}
>
  <div>
    <Component
      v={8}
    >
      <div>
        <Component
          v={11}
        >
          <div>
            11
          </div>
        </Component>
      </div>
    </Component>
  </div>
</Component>
`;

exports[`Summ the parent with a complex tree of children 1`] = `
<Component
  v={1}
>
  <div>
    <Component
      key="0"
      v={3}
    >
      <div>
        <Component
          v={6}
        >
          <div>
            6
          </div>
        </Component>
      </div>
    </Component>
    <Component
      key="1"
      v={5}
    >
      <div>
        5
      </div>
    </Component>
    <Component
      key="2"
      v={6}
    >
      <div>
        6
      </div>
    </Component>
  </div>
</Component>
`;

exports[`Summ the parent with its soley child 1`] = `
<Component
  v={2}
>
  <div>
    <Component
      v={8}
    >
      <div>
        8
      </div>
    </Component>
  </div>
</Component>
`;

exports[`Summ the parent with its two children separately 1`] = `
<Component
  v={1}
>
  <div>
    <Component
      key="0"
      v={3}
    >
      <div>
        3
      </div>
    </Component>
    <Component
      key="1"
      v={4}
    >
      <div>
        4
      </div>
    </Component>
  </div>
</Component>
`;
